### PR TITLE
planner: reimplement DEFAULT function to make the behavior consistent with MySQL when looking up the corresponding column (#19709)

### LINKED ISSUE #17807: P1-[4.0 bug hunting]-[SQL Plan Management]-Using the DEFAULT() function results in an ambiguous column error
Bug Hunter issue tidb-challenge-program/bug-hunting-issue#15

---

Consider the following statements:
```sql
CREATE TABLE t0(c0 INT DEFAULT 1);
SELECT t0.c0, t0.c0 FROM t0 ORDER BY DEFAULT(t0.c0); -- Column 'c0' in field list is ambiguous
```
Unexpectedly, executing the `SELECT` results in an error:
```
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 1
Server version: 5.7.25-TiDB- TiDB Server (Apache License 2.0), MySQL 5.7 compatible

Copyright (c) 2000, 2020, Oracle and/or its affiliates. All rights reserved.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> CREATE TABLE t0(c0 INT DEFAULT 1);
Query OK, 0 rows affected (0.00 sec)

mysql> SELECT t0.c0, t0.c0 FROM t0 ORDER BY DEFAULT(t0.c0);
ERROR 1052 (23000): Column 'c0' in field list is ambiguous
```
This is unexpected, since there is no ambiguity in the column reference (i.e., the columns are all fully qualified). It seems that this bug is caused by the `DEFAULT()` function; when removing it, the query works as expected.

When using MySQL 8.0.19 to execute this, the behavior is as expected:
```
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 18
Server version: 8.0.19 MySQL Community Server - GPL

Copyright (c) 2000, 2020, Oracle and/or its affiliates. All rights reserved.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> CREATE TABLE t0(c0 INT DEFAULT 1);
Query OK, 0 rows affected (0.04 sec)

mysql> SELECT t0.c0, t0.c0 FROM t0 ORDER BY DEFAULT(t0.c0);
Empty set (0.00 sec)
```
I found this based on the latest master (commit 7e71069ffa6052b497eaaaa2c3863bff4fe0ef3f) and can also reproduce it on the 4.0 RC:
```
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: 
Git Commit Hash: 
Git Branch: 
UTC Build Time: 2020-04-15 06:06:10
GoVersion: go1.13.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```
I'm not very sure how to classify an unexpected error, since such a case it not really accounted for in the classification rules. I think it might be P1 (The function and performance do not meet expectations.)  or P2 (Incompatible with MySQL).

### PR DESCRIPTION
**Pull request by ti-srebot**
cherry-pick #19709 to release-4.0

---

### What problem does this PR solve?

Issue Number: close #17807  

Problem Summary:

The previous Default function parsed to generate ast.ColumnName, and then further parsed the corresponding column. But in fact, the Default function does not need to obtain the value of the corresponding column from the storage, so it should not be further parsed by ast.ColumnName.

The previous implementation produced many incompatibility issues with MySQL, such as:
- #17807 
- `select b from t order by default(a)` will return Unknown column `a`, even if `t` has column `a`.

Have to merge https://github.com/pingcap/parser/pull/1027 first.

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:
- Modify the parser so that DEFAULT function is not generated ast.ColumnName.
- Add new field `allNames` for `PlanBuilder` and `expressionRewriter` to record the set of columns that can be used to find in `evalDefaultExpr`.
- Find the corresponding column in `evalDefaultExpr`. And frist, it will look for `allNames`. If it is not found, it will search in `expressionRewrite.names`.

### Check List 

Tests 

- Unit test

Side effects

### Release note 

- 
- reimplement DEFAULT function to make the behavior consistent with MySQL when looking up the corresponding column

### CODE REVIEW

**Comment by rebelice** [comment]
PTAL @qw4990 @lzmhhh123

File: planner/core/planbuilder.go [hunk 1]
```diff
@@ -448,6 +448,12 @@ type PlanBuilder struct {
 	buildingViewStack set.StringSet
 	// renamingViewName is the name of the view which is being renamed.
 	renamingViewName string
+	// CreateView needs this information to check whether exists nested view.
+	underlyingViewNames set.StringSet
```
Comment (qw4990) [inline planner/core/planbuilder.go:452]: This variable doesn't belong to https://github.com/pingcap/tidb/pull/19709, so please remove it.

**Review by ichn-hu — APPROVED** [review]
LGTM

**Review by qw4990 — APPROVED** [review]
LGTM

File: planner/core/planbuilder.go (hunk 1 above)
Comment (rebelice) [reply planner/core/planbuilder.go]: Done.

**Pull request merged**